src/coprocessor/memoset/mod.rs

```diff
@@ -0,0 +1,713 @@
+//! The `memoset` module implements a MemoSet.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use bellpepper_core::{boolean::Boolean, num::AllocatedNum, ConstraintSystem, SynthesisError};
+use itertools::Itertools;
+use once_cell::sync::OnceCell;
+
+use super::gadgets::construct_cons;
+use crate::circuit::gadgets::{
+    constraints::{enforce_equal, enforce_equal_zero, invert, sub},
+    pointer::AllocatedPtr,
+};
+use crate::field::LurkField;
+use crate::lem::circuit::GlobalAllocator;
+use crate::lem::Tag;
+use crate::lem::{pointers::Ptr, store::Store};
+use crate::tag::{ExprTag, Tag as XTag};
+use crate::z_ptr::ZPtr;
+
+use multiset::MultiSet;
+use query::{CircuitQuery, DemoQuery, Query};
+
+mod multiset;
+mod query;
+
+type ScopeQuery<F> = DemoQuery<F>;
+type ScopeCircuitQuery<F> = <DemoQuery<F> as Query<F>>::C;
+
+#[derive(Debug, Default)]
+pub struct Scope<F: LurkField, Q: Query<F>, M: MemoSet<F>> {
+    memoset: M,
+    // k => v
+    queries: HashMap<Ptr, Ptr>,
+    // k => ordered subqueries
+    dependencies: HashMap<Ptr, Vec<Q>>,
+    // kv pairs
+    toplevel_insertions: Vec<Ptr>,
+    // internally-inserted keys
+    internal_insertions: Vec<Ptr>,
+    // unique keys
+    all_insertions: Vec<Ptr>,
```

Comment:
These comments can be docstrings. Also, `Scope` would really benefit from documentation